pkg/instancetype/upgrade.go

```diff
@@ -0,0 +1,148 @@
+//nolint:lll
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+package instancetype
+
+import (
+	"context"
+
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	instancetypeapi "kubevirt.io/api/instancetype"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+)
+
+type Upgrader interface {
+	Upgrade(vm *virtv1.VirtualMachine) error
+}
+
+func (m *InstancetypeMethods) Upgrade(vm *virtv1.VirtualMachine) error {
```

Comment:
Can this method be called multiple times or only once?

Reply:
Yes it's idempotent, I can add some tests to assert this.

Reply:
https://github.com/kubevirt/kubevirt/pull/10545/commits/6bbc3fe190782d663a4d06eaecf37728b2528086 covers this btw, given the need to rebase I've kept it in a standalone commit for easier review but I'm happy to squash.